### ISSUE TITLE
Incorrect node/address used for mon

### ISSUE THREAD

**Issue by bootc** [rook/rook#5887]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Somehow the cluster ended up with the wrong IP address and node name for a mon in the `rook-ceph-mon-endpoints` ConfigMap. These were the deployed mons (slightly redacted):

```
rook-ceph-mon-f-6bc89786df-2d8zg   1/1     Running                 1          16d     192.168.32.41   node5.example.com
rook-ceph-mon-g-574fd9df4c-fmk2k   1/1     Running                 1          16d     192.168.32.43   node7.example.com
rook-ceph-mon-i-5ffb4bcc46-h5xpf   0/1     Init:CrashLoopBackOff   119        9h      192.168.32.32   node4.example.com
```

The `rook-ceph-mon-endpoints` ConfigMap at the time showed:

```
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["192.168.32.42:6789","192.168.32.32:6789","192.168.32.32:6789"]}]'
  data: f=192.168.32.42:6789,g=192.168.32.32:6789,i=192.168.32.32:6789
  mapping: '{"node":{"f":{"Name":"node6.example.com","Hostname":"node6.example.com","Address":"192.168.32.42"},"g":{"Name":"node4.example.com","Hostname":"node4.example.com","Address":"192.168.32.32"},"i":{"Name":"node4.example.com","Hostname":"node4.example.com","Address":"192.168.32.32"}}}'
```

Note in particular that mon `g` is running on `node7` with IP 192.168.32.43. The ConfigMap instead suggests it is on `node4` (with that node's IP address), which is the same as mon `i` which is failing to start.

Note that the duplicate IP address in the mon hosts list causes many invocations of the Ceph tools to crash, and this is what was causing mon `i` to fail to start. Editing the ConfigMap by hand to correct the details for mon `g`, and restarting the operator and mon `i` pods resolved the problem.

**Expected behavior:**
The `rook-ceph-mon-endpoints` ConfigMap should contain the correct node name and IP address.

**How to reproduce it (minimal and precise):**
I'm afraid I honestly don't know how to replicate this, but I have run into this issue on two different clusters.

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

I no longer have the logs from the time the issue started.

**Environment**:
* OS (e.g. from /etc/os-release): Debian Buster (10)
* Kernel (e.g. `uname -a`): Linux nodename 5.5.0-0.bpo.2-amd64 #1 SMP Debian 5.5.17-1~bpo10+1 (2020-04-23) x86_64 GNU/Linux
* Cloud provider or hardware configuration: Supermicro physical hardware
* Rook version (use `rook version` inside of a Rook Pod): 1.3.7 and 1.3.8*
* Storage backend version (e.g. for ceph do `ceph -v`): ceph/ceph:v15.2.4-20200630
* Kubernetes version (use `kubectl version`): 1.17.9
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): RKE
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)): HEALTH_WARN (109 daemons have recently crashed)

The issue has twice occurred while running Rook 1.3.7 with Ceph 15.2.4. My first futile attempt to fix this both times has been to upgrade the operator to 1.3.8 so at that point I lost the previous operator's logs, I'm afraid. The ceph crashes are uninteresting but look like this:

```
# ceph crash info 2020-07-23T11:18:38.510861Z_33d25d5d-a37f-4c76-9f0a-b43223b39b9e
{
    "assert_condition": "addr_mons.count(a) == 0",
    "assert_file": "/home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/15.2.4/rpm/el8/BUILD/ceph-15.2.4/src/mon/MonMap.h",
    "assert_func": "void MonMap::add(const mon_info_t&)",
    "assert_line": 199,
    "assert_msg": "/home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/15.2.4/rpm/el8/BUILD/ceph-15.2.4/src/mon/MonMap.h: In function 'void MonMap::add(const mon_info_t&)' thread 7f1a6b1d2700 time 2020-07-23T11:18:38.507986+0000\n/home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/15.2.4/rpm/el8/BUILD/ceph-15.2.4/src/mon/MonMap.h: 199: FAILED ceph_assert(addr_mons.count(a) == 0)\n",
    "assert_thread_name": "ceph-mon",
    "backtrace": [
        "(()+0x12dd0) [0x7f1a5fed1dd0]",
        "(gsignal()+0x10f) [0x7f1a5eb3a70f]",
        "(abort()+0x127) [0x7f1a5eb24b25]",
        "(ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1a9) [0x7f1a623cdb4d]",
        "(()+0x276d16) [0x7f1a623cdd16]",
        "(MonMap::add(mon_info_t const&)+0x78) [0x7f1a62755878]",
        "(MonMap::add(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, entity_addrvec_t const&, unsigned short, unsigned short)+0x12a) [0x7f1a62755aaa]",
        "(MonMap::init_with_ips(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x186) [0x7f1a62750606]",
        "(MonMap::build_initial(ceph::common::CephContext*, bool, std::ostream&)+0x2ca) [0x7f1a6275289a]",
        "(main()+0x94f) [0x55b0a08616cf]",
        "(__libc_start_main()+0xf3) [0x7f1a5eb266a3]",
        "(_start()+0x2e) [0x55b0a088fc6e]"
    ],
    "ceph_version": "15.2.4",
    "crash_id": "2020-07-23T11:18:38.510861Z_33d25d5d-a37f-4c76-9f0a-b43223b39b9e",
    "entity_name": "mon.i",
    "os_id": "centos",
    "os_name": "CentOS Linux",
    "os_version": "8 (Core)",
    "os_version_id": "8",
    "process_name": "ceph-mon",
    "stack_sig": "7d396511dc3f706ac7693db70f87bfa45895c2d3f7f3c80cd6ee58a818bb3c2a",
    "timestamp": "2020-07-23T11:18:38.510861Z",
    "utsname_hostname": "nodename",
    "utsname_machine": "x86_64",
    "utsname_release": "5.5.0-0.bpo.2-amd64",
    "utsname_sysname": "Linux",
    "utsname_version": "#1 SMP Debian 5.5.17-1~bpo10+1 (2020-04-23)"
}
```

This assertion failure seems to occur due to the duplicated IP address in the `mon_host` list.

**Comment by travisn** [comment]
@bootc This could be related to the issue just fixed by #5868 where a single mon ends up being deleted, then crashes continuously after the operator re-creates it due to the race condition described [here](https://github.com/rook/rook/pull/5869#issuecomment-662625143). However, in that case the mon log did not show a crash as you are seeing, the mon had a log message that it "committed suicide". 

If you see it again, definitely see if you can capture the operator log (and the previous operator log (`kubectl logs -p`) before it's restarted/upgraded.

**Comment by bootc** [comment]
I'm afraid this has happened again with 1.3.8 on the same cluster as above. Unfortunately once this occurs the operator starts crashing and by the time I noticed it it was on its 5th restart so the history leading up to the issue occurring had already been lost.

I have, however, not had a repeat of the issue so far on my other Rook/Ceph cluster that I upgraded to 1.3.9 and now recently to 1.4.0. I'll fix up this cluster shortly and upgrade it to 1.4.0 as well.

I'll try to get to it quicker if it happens again, else if I don't see it within a month or so with 1.4.0 I suspect you can indeed chalk it up to #5868 as you suggest; it does sound suspiciously like what might be happening.

**Comment by travisn** [comment]
Ok, shall we close this issue for now assuming it is fixed in 1.3.9 and newer since it's expected to be fixed by #5868? If you hit again we will reopen it.

**Comment by bootc** [comment]
Sure, works for me. Thanks for your time!
